File: android/app/src/test/java/com/booksync/ui/account/PasswordPolicyTest.kt
package com.booksync.ui.account

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Issue #205. These numbers are the Android third of a contract shared with
 * `server/schemas.py` and `web/src/lib/passwordPolicy.js`. Before this, the
 * sheet demanded 8 while the server and the web form allowed 6, so the three
 * surfaces disagreed about what a valid password was.
 */
class PasswordPolicyTest {

    @Test
    fun `policy is eight to one hundred twenty eight, matching the server`() {
        assertEquals(8, PasswordPolicy.MIN_LENGTH)
        assertEquals(128, PasswordPolicy.MAX_LENGTH)
    }

    @Test
    fun `message names both bounds`() {
        assertTrue(PasswordPolicy.MESSAGE.contains("8"))
        assertTrue(PasswordPolicy.MESSAGE.contains("128"))
    }

    @Test
    fun `one character under the floor is rejected`() {
        assertFalse(PasswordPolicy.isValid("a".repeat(PasswordPolicy.MIN_LENGTH - 1)))
    }

    @Test
    fun `exactly at the floor is accepted`() {
        assertTrue(PasswordPolicy.isValid("a".repeat(PasswordPolicy.MIN_LENGTH)))
    }

    @Test
    fun `exactly at the ceiling is accepted`() {
        assertTrue(PasswordPolicy.isValid("a".repeat(PasswordPolicy.MAX_LENGTH)))
    }

    @Test
    fun `one character over the ceiling is rejected`() {
        assertFalse(PasswordPolicy.isValid("a".repeat(PasswordPolicy.MAX_LENGTH + 1)))
    }

    @Test
    fun `an empty password is rejected`() {
        assertFalse(PasswordPolicy.isValid(""))
    }

    @Test
    fun `errorFor returns the message for a bad password and null for a good one`() {
        assertEquals(PasswordPolicy.MESSAGE, PasswordPolicy.errorFor("short"))
        assertNull(PasswordPolicy.errorFor("a".repeat(PasswordPolicy.MIN_LENGTH)))
    }
}

File: web/src/lib/passwordPolicy.test.js
import { describe, it, expect } from 'vitest'
import {
    PASSWORD_MAX_LENGTH,
    PASSWORD_MIN_LENGTH,
    PASSWORD_POLICY_MESSAGE,
    isPasswordValid,
    passwordError,
} from './passwordPolicy'

// Issue #205. These numbers are the web half of a three-way contract with
// server/schemas.py and android/.../PasswordPolicy.kt. Pinning them here means
// a drift shows up as a failing test rather than as a form that accepts what
// the API will 422.

describe('passwordPolicy', () => {
    it('is 8 to 128 characters, matching the server', () => {
        expect(PASSWORD_MIN_LENGTH).toBe(8)
        expect(PASSWORD_MAX_LENGTH).toBe(128)
    })

    it('names both bounds in the message', () => {
        expect(PASSWORD_POLICY_MESSAGE).toContain('8')
        expect(PASSWORD_POLICY_MESSAGE).toContain('128')
    })

    it('rejects one character under the floor and accepts the floor', () => {
        expect(isPasswordValid('a'.repeat(PASSWORD_MIN_LENGTH - 1))).toBe(false)
        expect(isPasswordValid('a'.repeat(PASSWORD_MIN_LENGTH))).toBe(true)
    })

    it('accepts the ceiling and rejects one character over it', () => {
        expect(isPasswordValid('a'.repeat(PASSWORD_MAX_LENGTH))).toBe(true)
        expect(isPasswordValid('a'.repeat(PASSWORD_MAX_LENGTH + 1))).toBe(false)
    })

    it('treats an empty or missing password as invalid', () => {
        expect(isPasswordValid('')).toBe(false)
        expect(isPasswordValid(undefined)).toBe(false)
        expect(isPasswordValid(null)).toBe(false)
    })

    it('returns the policy message for a bad password and null for a good one', () => {
        expect(passwordError('short')).toBe(PASSWORD_POLICY_MESSAGE)
        expect(passwordError('a'.repeat(PASSWORD_MIN_LENGTH))).toBeNull()
    })
})
